File: app/src/main/java/me/srikavin/quiz/repository/GameRepository.java
package me.srikavin.quiz.repository;

import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;

import androidx.annotation.NonNull;
import java9.util.stream.Collectors;
import java9.util.stream.StreamSupport;
import me.srikavin.quiz.model.AnswerResponse;
import me.srikavin.quiz.model.Quiz;
import me.srikavin.quiz.model.QuizAnswer;
import me.srikavin.quiz.model.QuizGameState;
import me.srikavin.quiz.model.QuizQuestion;

import static me.srikavin.quiz.MainActivity.TAG;

public enum GameRepository {
    INSTANCE;

    private LocalGameRepository localGameRepository = new LocalGameRepository();

    public void createGame(Quiz quiz, GameResponseHandler handler) {
        localGameRepository.createGame(quiz, handler);
    }

    public void submitAnswer(String id, QuizAnswer answer) {
        localGameRepository.submitAnswer(id, answer);
    }

    public enum ErrorCodes {
        UNKNOWN_ERROR(0),
        NETWORK_ERROR(5),
        SERVER_ERROR(6);

        private int code;

        ErrorCodes(int code) {
            this.code = code;
        }

        static ErrorCodes fromCode(int errorCode) {
            for (ErrorCodes e : values()) {
                if (errorCode == e.code) {
                    return e;
                }
            }
            return UNKNOWN_ERROR;
        }
    }

    interface GameService {
        void createGame(Quiz quiz, GameResponseHandler handler);

        void submitAnswer(String id, QuizAnswer answer);
    }

    public abstract static class GameResponseHandler {
        public void handleAnswer(AnswerResponse response) {
            //By default, do nothing
        }

        public void handleQuestion(QuizQuestion question) {
            //By default, do nothing
        }

        public void handleGameCreate(String id, GameInfo info) {
            //By default, do nothing
        }

        public void handleGameStats(GameStats stats) {

        }

        public void handleScoreChange(int score) {

        }

        public void handleGameStateChange(QuizGameState state) {

        }

        public void handleGameTimeChange(int timeLeft) {

        }

        public void handleErrors(@NonNull ErrorCodes... errors) {
            //By default, print error codes
            for (GameRepository.ErrorCodes e : errors) {
                Log.w(TAG, "Ignored error code: " + e.name());
            }
        }
    }

    public static class GameStats {
        public final int correct;
        public final List<QuizAnswer> chosen;
        public final List<QuizQuestion> quizQuestions;
        public final double percentCorrect;

        public GameStats(int correct, List<QuizAnswer> chosen, List<QuizQuestion> quizQuestions) {
            this.correct = correct;
            this.chosen = chosen;
            this.quizQuestions = quizQuestions;
            percentCorrect = correct / (double) quizQuestions.size();
        }
    }

    public static class GameInfo {
        public final int timePerQuestion;
        public final int numberOfQuestions;

        public GameInfo(int timePerQuestion, int numberOfQuestions) {
            this.timePerQuestion = timePerQuestion;
            this.numberOfQuestions = numberOfQuestions;
        }
    }

    public static class LocalGameRepository implements GameService {
        private Map<String, Game> idGameMap = new HashMap<>();

        @Override
        public void createGame(Quiz quiz, GameResponseHandler handler) {
            String id = UUID.randomUUID().toString();
            Game game = new Game(id, quiz, handler);
            idGameMap.put(id, game);
            game.setState(QuizGameState.WAITING_FOR_PLAYERS);
            handler.handleGameCreate(id, new GameInfo(30, quiz.questions.size()));
            game.setState(QuizGameState.IN_PROGRESS);
            nextQuestion(game);
        }

        @Override
        public void submitAnswer(String id, QuizAnswer answer) {
            Game game = idGameMap.get(id);
            assert game != null;

            if (game.waitingForAnswer || game.state == QuizGameState.FINISHED) {
                return;
            }

            boolean correct = answer != null && answer.correct;

            game.chosen.add(answer);

            if (correct) {
                game.correct++;
                game.score += 150;
                game.handler.handleScoreChange(game.score);
            }

            QuizQuestion cur = game.quiz.questions.get(game.currentQuestion);
            List<QuizAnswer> correctAnswers = StreamSupport.stream(cur.answers)
                    .filter(e -> e.correct).collect(Collectors.toList());

            game.handler.handleAnswer(new AnswerResponse(correct, cur, correctAnswers));

            game.waitingForAnswer = true;
            // Delay the passage of the next question to allow for animations and effects
            game.countDownTimer.schedule(new TimerTask() {
                @Override
                public void run() {
                    nextQuestion(game);
                    game.waitingForAnswer = false;
                }
            }, 3500);
        }

        private void nextQuestion(Game game) {
            game.currentQuestion++;
            if (game.currentQuestion < game.total) {
                game.setCurrentQuestion(game.currentQuestion);
                QuizQuestion cur = game.quiz.questions.get(game.currentQuestion);
                Collections.shuffle(cur.answers);
                game.handler.handleQuestion(cur);
                if (game.executingTask != null) {
                    game.executingTask.cancel();
                }
                game.executingTask = new TimerTask() {
                    int seconds = 31;

                    @Override
                    public void run() {
                        seconds--;
                        game.handler.handleGameTimeChange(seconds);
                        if (seconds <= 0) {
                            submitAnswer(game.id, null);
                            cancel();
                        }
                    }
                };
                game.countDownTimer.scheduleAtFixedRate(game.executingTask, 0, 1000);
            } else {
                game.currentQuestion--;
                game.setState(QuizGameState.FINISHED);
                game.handler.handleGameStats(new GameStats(game.correct, game.chosen, game.quiz.questions));
                game.countDownTimer.cancel();
            }

        }

        class Game {
            Quiz quiz;
            String id;
            int total;
            int score;
            int currentQuestion = -1;
            int correct = 0;
            boolean waitingForAnswer = false;
            List<QuizAnswer> chosen = new ArrayList<>();
            GameResponseHandler handler;
            Timer countDownTimer = new Timer();
            TimerTask executingTask;
            private QuizGameState state;

            public Game(String id, Quiz quiz, GameResponseHandler handler) {
                this.id = id;
                this.handler = handler;
                this.quiz = quiz;
                Collections.shuffle(quiz.questions);
                total = quiz.questions.size();
            }

            void setState(QuizGameState state) {
                this.state = state;
                handler.handleGameStateChange(state);
            }

            void setCurrentQuestion(int currentQuestion) {
                this.currentQuestion = currentQuestion;
                state.setCurrentQuestion(currentQuestion);
                this.setState(state);
            }

        }
    }
}

File: app/src/main/java/me/srikavin/quiz/repository/GameRepository.kt
package me.srikavin.quiz.repository

import android.util.Log
import me.srikavin.quiz.model.AnswerResponse
import me.srikavin.quiz.model.Quiz
import me.srikavin.quiz.model.QuizGameState
import me.srikavin.quiz.network.common.model.data.QuizAnswerModel
import me.srikavin.quiz.network.common.model.data.QuizQuestionModel
import me.srikavin.quiz.repository.local.LocalGameRepository
import me.srikavin.quiz.view.main.TAG
import java.io.Serializable

/**
 * A wrapper class that abstracts the data type of a GameId and ensured typesafety
 */
inline class GameID(
        /**
         * The actual game id stored as a string
         */
        val value: String
)

/**
 * A repository that handles creating games and communicating events to and from the app and the game
 * server. Supports remote and local servers.
 */
object GameRepository {
    private val localGameRepository = LocalGameRepository()
    private val remoteGameRepository = RemoteGameRepository()

    /**
     * Creates a game with the given quiz. If [remote] is set to `true`, the game is created on a
     * remote game server.
     */
    fun createGame(quiz: Quiz, handler: GameResponseHandler, remote: Boolean) {
        if (remote) {
            remoteGameRepository.createGame(quiz, handler)
        } else {
            localGameRepository.createGame(quiz, handler)
        }
    }

    /**
     * Submits an answer to the underlying game with the given game id.
     */
    fun submitAnswer(id: GameID, answer: QuizAnswerModel) {
        if (localGameRepository.hasGame(id)) {
            localGameRepository.submitAnswer(id, answer)
        }
        if (remoteGameRepository.hasGame(id)) {
            remoteGameRepository.submitAnswer(id, answer)
        }
    }

    /**
     * Quits the given game. Rejoining is not possible.
     */
    fun quit(id: GameID) {
        if (localGameRepository.hasGame(id)) {
            localGameRepository.quit(id)
        }
        if (remoteGameRepository.hasGame(id)) {
            remoteGameRepository.quit(id)
        }
    }

    /**
     * Sends a signal to the backing repository to stop matchmaking
     * If offline, this has no effect
     */
    fun stopMatchmaking() {
        localGameRepository.stopMatchmaking()
    }

    /**
     * A enumeration of possible errors that may occur when using this repository
     */
    enum class ErrorCodes constructor(private val code: Int) {
        /**
         * Indicates an unknown error occurred
         */
        UNKNOWN_ERROR(0),
        /**
         * Indicates a network error occurred
         */
        NETWORK_ERROR(5),
        /**
         * Indicates a server error occurred
         */
        SERVER_ERROR(6);


        companion object {
            internal fun fromCode(errorCode: Int): ErrorCodes {
                for (e in values()) {
                    if (errorCode == e.code) {
                        return e
                    }
                }
                return UNKNOWN_ERROR
            }
        }
    }

    internal interface GameService {
        fun stopMatchmaking()

        fun createGame(quiz: Quiz, handler: GameResponseHandler)

        fun submitAnswer(id: GameID, answer: QuizAnswerModel?)

        fun hasGame(id: GameID): Boolean

        fun quit(id: GameID)
    }

    /**
     * An interface that should be used to handle events dispatched by a [GameRepository]. Not all
     * methods need to be overridden.
     */
    abstract class GameResponseHandler {
        /**
         * Handler for displaying the correct answer
         */
        open fun handleAnswer(response: AnswerResponse) {
            //By default, do nothing
        }

        /**
         * Handler for a new question or a change in current question
         */
        open fun handleQuestion(question: QuizQuestionModel) {
            //By default, do nothing
        }


        /**
         * Handler for updates to the GameInfo
         */
        open fun handleGameInfo(info: GameRepository.GameInfo) {

        }


        /**
         * Handler called when a new game is created with the game's id
         */
        open fun handleGameCreate(id: GameID) {

        }

        /**
         * Handler for Game Stats update
         */
        open fun handleGameStats(stats: GameStats) {

        }

        /**
         * Handler for new score update
         */
        open fun handleScoreChange(score: Int) {

        }

        /**
         * Handler for changes to the game state
         */
        open fun handleGameStateChange(state: QuizGameState) {

        }

        /**
         * Handler for changes to the time remaining for the current question
         */
        open fun handleGameTimeChange(timeLeft: Int) {

        }

        /**
         * Called when an exception occurs. These will be categorized into [ErrorCodes]. Unknown errors
         * will be returned as [ErrorCodes.UNKNOWN_ERROR].
         */
        fun handleErrors(vararg errors: ErrorCodes) {
            //By default, print error codes
            for (e in errors) {
                Log.w(TAG, "Ignored error code: " + e.name)
            }
        }
    }

    /**
     * Data class for holding the final game statistics
     */
    class GameStats(
            /**
             * The number of correct answers
             */
            val correct: Int,
            /**
             * The final score of the game player
             */
            val score: Int,
            /**
             * A list of answers chosen by the player. Null is given if no answer was chosen/
             */
            val chosen: List<QuizAnswerModel?>,
            /**
             * A list of questions present in the quiz
             */
            val quizQuestions: List<QuizQuestionModel>
    ) : Serializable {
        /**
         * The total number of quiz questions present in the quiz
         */
        val total: Int = quizQuestions.size
        /**
         * The percentage of questions the quiz player answered correctly
         */
        val percentCorrect: Double

        init {
            percentCorrect = if (total != 0) {
                correct / total.toDouble()
            } else {
                0.0
            }
        }
    }

    /**
     * A data class containing information about the quiz-game itself
     */
    class GameInfo(
            /**
             * The number of time taken for each question
             */
            val timePerQuestion: Int,
            /**
             * The number of questions present in this quiz
             */
            val numberOfQuestions: Int
    )

}
